### DISCUSSION TITLE
PIXIJS Graphic blend,Color problem of superimposed part

### DISCUSSION THREAD

**Discussion by pcbren1** [pixijs/pixijs#7573] [Q&A]
![11123](https://user-images.githubusercontent.com/86771584/124263839-5ab6ab00-db66-11eb-8ad7-e247b300511b.png)

**Comment by ivanpopelyshev** [comment]
Which blendMode do you want to implement? if its just Overlay - you can use filter from https://github.com/pixijs/pixi-picture .

**Reply by ivanpopelyshev** [reply]
It looks like DIFFERENCE? If you want this thing, you can either extend shaders in @pixi/picture or request it in the repo, i can add it.

**Reply by pcbren1** [reply]
![A1](https://user-images.githubusercontent.com/86771584/124270737-2dbac600-db6f-11eb-9183-ddc8435e7e81.png)
![A2](https://user-images.githubusercontent.com/86771584/124270749-30b5b680-db6f-11eb-8f52-0443cad1e9b1.png)

**Reply by pcbren1** [reply]
blendMode is a color overlay algorithm. How to rewrite this algorithm. Add PIXI.BLEND_MODES of modes

**Reply by ivanpopelyshev** [reply]
if you know constants for Porter-Duff, you can specify it, just make sure number doesnt collide with other pixi blends in https://github.com/pixijs/pixijs/blob/4d92e01380417d10e80624dfa5360bfe0708628c/packages/constants/src/index.ts#L95-L129

```js
const gl = renderer.context.gl;
renderer.state.blendModes[myBlendModeNumber] = [gl.ONE, gl.ONE] ; 
```

Here's how other blend modes are set up : https://github.com/pixijs/pixijs/blob/dev/packages/core/src/state/utils/mapWebGLBlendModesToPixi.ts

If blending mode is not possible to make with Porter-Duff and premultiplied alpha, you have to add it to https://github.com/pixijs/pixi-picture :
write a part like https://github.com/pixijs/pixi-picture/blob/aed834cbf877148ad4f9d485f0d23b16b2708784/src/ShaderParts.ts#L20-L40, 

then make a full version and put it in blendFullArray
https://github.com/pixijs/pixi-picture/blob/aed834cbf877148ad4f9d485f0d23b16b2708784/src/ShaderParts.ts#L127-L129

necessary things are exposed in `@pixi/picture` module, so you dont need to rebuild it, you can just do it from outside. Then you can use `getBlendFilter(myBlendModeNumber)` and put it in filters for your object.

In any case, please read source code of files i mentioned.

**Reply by ivanpopelyshev** [reply]
Formulas for common blending operations are here: https://www.w3.org/TR/compositing-1/#blending

For example, difference is just one line `vec3 B = abs(Cb - Cs);`
